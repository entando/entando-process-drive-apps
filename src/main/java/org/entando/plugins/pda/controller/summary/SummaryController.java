package org.entando.plugins.pda.controller.summary;

import static org.entando.plugins.pda.controller.AuthPermissions.SUMMARY_DATA_REPOSITORY_LIST;
import static org.entando.plugins.pda.controller.AuthPermissions.SUMMARY_GET;
import static org.entando.plugins.pda.controller.AuthPermissions.SUMMARY_TYPE_LIST;
import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;

import io.swagger.annotations.Api;
import io.swagger.annotations.ApiOperation;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.entando.plugins.pda.core.engine.Connection;
import org.entando.plugins.pda.core.model.summary.Summary;
import org.entando.plugins.pda.core.service.summary.DataService;
import org.entando.plugins.pda.core.service.summary.SummaryService;
import org.entando.plugins.pda.service.ConnectionService;
import org.entando.web.response.SimpleRestResponse;
import org.springframework.security.access.annotation.Secured;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@Api(tags = "Summary")
@RequestMapping(path = "/connections/{connId}/summaries")
@RequiredArgsConstructor
public class SummaryController {

    private final ConnectionService connectionService;
    private final SummaryService summaryService;
    private final DataService dataService;

    @Secured(SUMMARY_DATA_REPOSITORY_LIST)
    @ApiOperation(notes = "Lists all Summary Data Repositories for a Connection",
            nickname = "listSummaryDataRepositories", value = "LIST DataRepository")
    @GetMapping(value = "/repositories", produces = APPLICATION_JSON_VALUE)
    public SimpleRestResponse<List<String>> listSummaryDataRepositories(@PathVariable String connId) {
        Connection connection = connectionService.get(connId);
        return new SimpleRestResponse<>(dataService.listDataRepositories(connection.getEngine()));
    }

    @Secured(SUMMARY_TYPE_LIST)
    @ApiOperation(notes = "Lists all Summary Types for a given connection", nickname = "listSummaryTypes",
            value = "LIST SummaryType")
    @GetMapping(value = "/summaryTypes", produces = APPLICATION_JSON_VALUE)
    public SimpleRestResponse<List<String>> listSummaryTypes(@PathVariable String connId) {
        return new SimpleRestResponse<>(summaryService.listSummaryTypes());
    }

    @Secured(SUMMARY_GET)
    @ApiOperation(notes = "Gets the calculated Summary",
            nickname = "calculateSummary", value = "GET Summary")
    @PostMapping(value = "/summaryTypes/{summaryType}", produces = APPLICATION_JSON_VALUE)
    public SimpleRestResponse<Summary> calculateSummary(@PathVariable String connId, @PathVariable String summaryType,
            @RequestBody String request) {
        Connection connection = connectionService.get(connId);
        return new SimpleRestResponse<>(summaryService.getSummaryProcessor(summaryType)
                .getSummary(connection, request));
    }
}
