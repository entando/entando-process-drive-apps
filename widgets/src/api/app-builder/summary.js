import { SERVICE, METHODS } from 'api/constants';
import mockedSummary from 'mocks/summaries/summary';

import makeRequest from 'api/makeRequest';

export const getSummary = async (connection, summaryId, params = '?frequency=monthly') => {
  return makeRequest({
    domain: SERVICE.URL,
    uri: `/connections/${connection}/summaries/${summaryId}${params}`,
    method: METHODS.GET,
    mockResponse: mockedSummary,
    useAuthentication: false,
  });
};
