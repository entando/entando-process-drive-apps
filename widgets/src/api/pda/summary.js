import makeRequest from 'api/makeRequest';
import { DOMAINS, METHODS } from 'api/constants';
import MOCK_SUMMARIES from 'mocks/summary/summaries';
import mockedSummary from 'mocks/summary/summary';

export const getSummaries = async connection =>
  makeRequest({
    domain: DOMAINS.PDA,
    uri: `/connections/${connection}/summaries`,
    method: METHODS.GET,
    mockResponse: MOCK_SUMMARIES,
    useAuthentication: false,
  });

export const getSummary = async (connection, summaryId, frequency = 'monthly') =>
  makeRequest({
    domain: DOMAINS.PDA,
    uri: `/connections/${connection}/summaries/${summaryId}?frequency=${frequency}`,
    method: METHODS.GET,
    mockResponse: mockedSummary,
    useAuthentication: false,
  });
