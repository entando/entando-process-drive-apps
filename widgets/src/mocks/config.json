{
  "payload": {
    "code": "pda_task_list",
    "config": {
      "groups": "[{\"label\":\"Administrators\",\"key\":\"Administrators\",\"checked\":true},{\"label\":\"broker\",\"key\":\"broker\",\"checked\":true},{\"label\":\"admin\",\"key\":\"admin\",\"checked\":false},{\"label\":\"rest-all\",\"key\":\"rest-all\",\"checked\":false},{\"label\":\"kiemgmt\",\"key\":\"kiemgmt\",\"checked\":true},{\"label\":\"kie-server\",\"key\":\"kie-server\",\"checked\":true},{\"label\":\"appraiser\",\"key\":\"appraiser\",\"checked\":false},{\"label\":\"manager\",\"key\":\"manager\",\"checked\":true},{\"label\":\"supplier\",\"key\":\"supplier\",\"checked\":true},{\"label\":\"IT\",\"key\":\"IT\",\"checked\":true},{\"label\":\"PM\",\"key\":\"PM\",\"checked\":true}]",
      "options": "[{\"key\":\"newPageOnClick\",\"label\":\"Open new page on table row click\",\"checked\":false},{\"key\":\"showClaim\",\"label\":\"Show Claim Button\",\"checked\":true},{\"key\":\"showComplete\",\"label\":\"Show Complete Button\",\"checked\":true}]",
      "columns": "[{\"name\":\"task-is-skipable\",\"position\":0,\"isVisible\":false},{\"name\":\"task-parent-id\",\"position\":1,\"isVisible\":false},{\"name\":\"task-created-by\",\"position\":2,\"isVisible\":true},{\"name\":\"task-input-data.application.property.saleprice\",\"position\":3,\"isVisible\":true},{\"name\":\"task-input-data.TaskName\",\"position\":4,\"isVisible\":true},{\"name\":\"initiator\",\"position\":5,\"isVisible\":true},{\"name\":\"task-created-on\",\"position\":6,\"isVisible\":true},{\"name\":\"task-priority\",\"position\":7,\"isVisible\":false},{\"name\":\"task-proc-def-id\",\"position\":8,\"isVisible\":true},{\"name\":\"task-form\",\"position\":9,\"isVisible\":true},{\"name\":\"task-skippable\",\"position\":10,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.name\",\"position\":11,\"isVisible\":true},{\"name\":\"task-actual-owner\",\"position\":12,\"isVisible\":true},{\"name\":\"task-activation-time\",\"position\":13,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.address\",\"position\":14,\"isVisible\":true},{\"name\":\"task-input-data.application.errors\",\"position\":15,\"isVisible\":true},{\"name\":\"task-name\",\"position\":16,\"isVisible\":true},{\"name\":\"task-input-data.application.property.address\",\"position\":17,\"isVisible\":true},{\"name\":\"task-subject\",\"position\":18,\"isVisible\":true},{\"name\":\"containerId\",\"position\":19,\"isVisible\":true},{\"name\":\"task-container-id\",\"position\":20,\"isVisible\":true},{\"name\":\"processDefinitionId\",\"position\":21,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.ssn\",\"position\":22,\"isVisible\":true},{\"name\":\"task-description\",\"position\":23,\"isVisible\":true},{\"name\":\"task-input-data.Skippable\",\"position\":24,\"isVisible\":true},{\"name\":\"task-workitem-id\",\"position\":25,\"isVisible\":true},{\"name\":\"task-expiration-time\",\"position\":26,\"isVisible\":true},{\"name\":\"task-input-data.application.property.age\",\"position\":27,\"isVisible\":true},{\"name\":\"task-status\",\"position\":28,\"isVisible\":true},{\"name\":\"task-input-data.application.mortgageamount\",\"position\":29,\"isVisible\":true},{\"name\":\"task-process-instance-id\",\"position\":30,\"isVisible\":true},{\"name\":\"task-input-data.ActorId\",\"position\":31,\"isVisible\":true},{\"name\":\"task-process-id\",\"position\":32,\"isVisible\":true},{\"name\":\"task-input-data.application.property.locale\",\"position\":33,\"isVisible\":true},{\"name\":\"task-input-data.application.amortization\",\"position\":34,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.creditrating\",\"position\":35,\"isVisible\":true},{\"name\":\"application\",\"position\":36,\"isVisible\":true},{\"name\":\"task-input-data.NodeName\",\"position\":37,\"isVisible\":true},{\"name\":\"task-proc-inst-id\",\"position\":38,\"isVisible\":true},{\"name\":\"task-input-data.application.downpayment\",\"position\":39,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.annualincome\",\"position\":40,\"isVisible\":true},{\"name\":\"task-id\",\"position\":41,\"isVisible\":true}]",
      "knowledgeSource": "kieStaging",
      "process": "itorders.orderhardware@itorders_1.0.0-SNAPSHOT"
    }
  },
  "metaData": {
    "status": "draft"
  },
  "errors": []
}
