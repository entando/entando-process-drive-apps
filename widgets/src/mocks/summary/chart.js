export default {
  payload: {
    series: [
      {
        dataType: 'requests',
        values: [
          {
            date: '2020-01-01',
            value: 18.0,
          },
          {
            date: '2019-01-01',
            value: 111.0,
          },
          {
            date: '2018-01-01',
            value: 0.0,
          },
          {
            date: '2017-01-01',
            value: 0.0,
          },
          {
            date: '2016-01-01',
            value: 0.0,
          },
        ],
        card: {
          value: 18.0,
          percentage: 0.16216216216216217,
        },
      },
      {
        dataType: 'cases',
        values: [
          {
            date: '2020-01-01',
            value: 18.0,
          },
          {
            date: '2019-01-01',
            value: 111.0,
          },
          {
            date: '2018-01-01',
            value: 0.0,
          },
          {
            date: '2017-01-01',
            value: 0.0,
          },
          {
            date: '2016-01-01',
            value: 0.0,
          },
        ],
        card: {
          value: 18.0,
          percentage: 0.16216216216216217,
        },
      },
    ],
  },
  errors: [],
};
