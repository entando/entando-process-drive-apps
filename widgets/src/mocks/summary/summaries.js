export default {
  payload: [
    {
      description: 'Requests',
      engine: 'pam',
      id: 'requests',
    },
  ],
  errors: [],
};
