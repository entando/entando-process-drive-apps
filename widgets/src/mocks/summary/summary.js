export const MOCK_SUMMARY_TYPES = {
  payload: ['requests', 'impressions'],
  errors: [],
};

export const MOCK_SUMMARY_CARD = {
  payload: {
    value: 2123,
    percentage: 0.9243,
  },
  errors: [],
};

export const MOCK_SUMMARY_CHART = {
  payload: {
    series: [
      {
        dataType: 'requests',
        values: [
          {
            date: '2020-01-01',
            value: 18.0,
          },
          {
            date: '2019-01-01',
            value: 111.0,
          },
          {
            date: '2018-01-01',
            value: 0.0,
          },
          {
            date: '2017-01-01',
            value: 0.0,
          },
          {
            date: '2016-01-01',
            value: 0.0,
          },
        ],
        card: {
          value: 18.0,
          percentage: 0.16216216216216217,
        },
      },
      {
        dataType: 'cases',
        values: [
          {
            date: '2020-01-01',
            value: 18.0,
          },
          {
            date: '2019-01-01',
            value: 111.0,
          },
          {
            date: '2018-01-01',
            value: 0.0,
          },
          {
            date: '2017-01-01',
            value: 0.0,
          },
          {
            date: '2016-01-01',
            value: 0.0,
          },
        ],
        card: {
          value: 18.0,
          percentage: 0.16216216216216217,
        },
      },
    ],
  },
  errors: [],
};

export const getMockSummaryOf = type => {
  switch (type) {
    case 'Card':
      return MOCK_SUMMARY_CARD;
    case 'Chart':
      return MOCK_SUMMARY_CHART;
    default:
      return {};
  }
};

export default getMockSummaryOf;
