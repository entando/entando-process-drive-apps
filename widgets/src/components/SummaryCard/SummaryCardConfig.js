/* eslint-disable no-console */
import React from 'react';
import PropTypes from 'prop-types';
import {
  FormGroup,
  ControlLabel,
  FormControl,
  Button,
  HelpBlock,
  Row,
  Col,
} from 'patternfly-react';

import { getConnections } from 'api/pda/connections';
import { getPageWidget, putPageWidget } from 'api/app-builder/pages';

import 'patternfly-react/dist/css/patternfly-react.css';
import 'patternfly/dist/css/patternfly.css';
import 'patternfly/dist/css/patternfly-additions.css';

class SummaryCardConfig extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      sourceList: [],
      settings: '{"summaryId": ""}',
      knowledgeSource: '',
    };

    this.onChangeKnowledgeSource = this.onChangeKnowledgeSource.bind(this);
    this.onChangeSettings = this.onChangeSettings.bind(this);
    this.handleSave = this.handleSave.bind(this);
  }

  async componentDidMount() {
    const { frameId, pageCode } = this.props;

    // getting list of Kie server connections
    const sourceList = await getConnections();
    this.setState({ sourceList: sourceList.payload });

    // getting existing configs
    const pageWidgetsConfigs = await getPageWidget(pageCode, frameId, 'SUMMARY_CARD');

    const configs = pageWidgetsConfigs.payload && pageWidgetsConfigs.payload.config;

    if (configs && configs.knowledgeSource) {
      this.onChangeKnowledgeSource(configs.knowledgeSource, () => {
        if (configs.settings) {
          this.setState({
            settings: configs.settings,
          });
        }
      });
    }
  }

  onChangeKnowledgeSource(e, cb = () => {}) {
    const knowledgeSource = e.target ? e.target.value : e;
    this.setState({ knowledgeSource }, cb);
  }

  onChangeSettings({ target: { value } }) {
    this.setState({ settings: value });
  }

  async handleSave() {
    const { frameId, pageCode, widgetCode } = this.props;
    const { knowledgeSource, settings } = this.state;

    const body = JSON.stringify({
      code: widgetCode,
      config: {
        knowledgeSource,
        settings,
      },
    });

    try {
      const response = await putPageWidget(pageCode, frameId, body);
      console.log('Configs got saved', response);
    } catch (error) {
      console.log('Error while saving configs', error);
    }
  }

  render() {
    const { knowledgeSource, sourceList, settings } = this.state;

    return (
      <div>
        <form>
          <Row>
            <Col xs={12}>
              <FormGroup controlId="connection">
                <ControlLabel>Knowledge Source</ControlLabel>
                <select
                  className="form-control"
                  value={knowledgeSource}
                  onChange={this.onChangeKnowledgeSource}
                >
                  <option value="">Select...</option>
                  {sourceList.map(source => (
                    <option key={source.name} value={source.name}>
                      {source.name}
                    </option>
                  ))}
                </select>
                <HelpBlock>Select one of the Kie server connections.</HelpBlock>
              </FormGroup>
            </Col>
          </Row>
          {knowledgeSource && (
            <section>
              <legend>Settings</legend>
              <Row>
                <Col xs={12}>
                  <FormGroup bsClass="form-group" controlId="textarea">
                    <ControlLabel bsClass="control-label">Config</ControlLabel>
                    <FormControl
                      bsClass="form-control"
                      componentClass="textarea"
                      value={settings}
                      onChange={this.onChangeSettings}
                    />
                  </FormGroup>
                </Col>
              </Row>
              <Row>
                <Col xs={12} className="text-right">
                  <Button bsClass="btn" bsStyle="primary" onClick={this.handleSave}>
                    Save
                  </Button>
                </Col>
              </Row>
            </section>
          )}
        </form>
      </div>
    );
  }
}

SummaryCardConfig.propTypes = {
  frameId: PropTypes.string.isRequired,
  widgetCode: PropTypes.string.isRequired,
  pageCode: PropTypes.string.isRequired,
};

export default SummaryCardConfig;
