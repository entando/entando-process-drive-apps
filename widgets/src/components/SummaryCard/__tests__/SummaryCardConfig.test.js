import { render, wait } from '@testing-library/react';
import React from 'react';
import 'mocks/i18nMock';

import PAGE_CONFIG from 'mocks/app-builder/pages';
import { MOCK_SUMMARY_TYPES } from 'mocks/summary/summary';
import CONNECTIONS from 'mocks/pda/connections';
import SummaryCardConfig from 'components/SummaryCard/SummaryCardConfig';

describe('<SummaryCardConfig />', () => {
  it('renders snapshot correctly', async () => {
    fetch
      .once(JSON.stringify(CONNECTIONS))
      .once(JSON.stringify(PAGE_CONFIG.SUMMARY_CARD))
      .once(JSON.stringify(MOCK_SUMMARY_TYPES));

    const { container } = render(<SummaryCardConfig config={{}} />);

    await wait(() => expect(container).toMatchSnapshot());
  });
});
