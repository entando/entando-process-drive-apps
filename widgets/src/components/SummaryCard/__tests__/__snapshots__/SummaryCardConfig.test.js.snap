// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SummaryCardConfig /> renders snapshot correctly 1`] = `
<div>
  <div>
    <form>
      <div
        class="row"
      >
        <div
          class="col-xs-12"
        >
          <div
            class="form-group"
          >
            <label
              class="control-label"
              for="connection"
            >
              Knowledge Source
            </label>
            <select
              class="form-control"
            >
              <option
                value=""
              >
                Select...
              </option>
              <option
                value="kieStaging"
              >
                kieStaging
              </option>
              <option
                value="kie2Server"
              >
                kie2Server
              </option>
              <option
                value="test"
              >
                test
              </option>
            </select>
            <span
              class="help-block"
            >
              Select one of the Kie server connections.
            </span>
          </div>
        </div>
      </div>
      <section>
        <legend>
          Settings
        </legend>
        <div
          class="row"
        >
          <div
            class="col-xs-12"
          >
            <div
              class="form-group"
            >
              <label
                class="control-label"
                for="textarea"
              >
                Config
              </label>
              <textarea
                class="form-control"
                id="textarea"
              >
                {"summaryId": "request"}
              </textarea>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="text-right col-xs-12"
          >
            <button
              class="btn btn-primary"
              type="button"
            >
              Save
            </button>
          </div>
        </div>
      </section>
    </form>
  </div>
</div>
`;
